crates/manifest/src/schema/v2.rs

```diff
@@ -82,12 +87,123 @@ pub struct Trigger {
 }
 
 /// One or many `ComponentSpec`(s)
-#[derive(Clone, Debug, Serialize, Deserialize)]
+#[derive(Clone, Debug, Serialize, Deserialize, JsonSchema)]
 #[serde(transparent)]
-pub struct OneOrManyComponentSpecs(#[serde(with = "one_or_many")] pub Vec<ComponentSpec>);
+pub struct OneOrManyComponentSpecs(
+    #[serde(with = "one_or_many")]
+    #[schemars(schema_with = "one_or_many_schema::<ComponentSpec>")]
+    pub Vec<ComponentSpec>,
+);
+
+fn one_or_many_schema<T: schemars::JsonSchema>(
+    gen: &mut schemars::gen::SchemaGenerator,
+) -> schemars::schema::Schema {
+    schemars::schema::Schema::Object(schemars::schema::SchemaObject {
+        subschemas: Some(Box::new(schemars::schema::SubschemaValidation {
+            one_of: Some(vec![
+                gen.subschema_for::<T>(),
+                gen.subschema_for::<Vec<T>>(),
+            ]),
+            ..Default::default()
+        })),
+        ..Default::default()
+    })
+}
+
+fn toml_table_schema(_gen: &mut schemars::gen::SchemaGenerator) -> schemars::schema::Schema {
+    schemars::schema::Schema::Object(schemars::schema::SchemaObject {
+        instance_type: Some(schemars::schema::SingleOrVec::Single(Box::new(
+            schemars::schema::InstanceType::Object,
+        ))),
+        ..Default::default()
+    })
+}
+
+fn map_of_toml_tables_schema(
+    _gen: &mut schemars::gen::SchemaGenerator,
+) -> schemars::schema::Schema {
+    schemars::schema::Schema::Object(schemars::schema::SchemaObject {
+        instance_type: Some(schemars::schema::SingleOrVec::Single(Box::new(
+            schemars::schema::InstanceType::Object,
+        ))),
+        ..Default::default()
+    })
+}
```

Comment:
Perhaps these should move into the `mod` below? (with their `_schema` tails removed?)